### PR TITLE
CryptoAlg-574: Implement ARMv8 PerlAsm of beeu_mod_inverse_vartime

### PR DESCRIPTION
**Pull request by nebeid**
### Issues:
CryptoAlg-574

### Description of changes: 
Continuing on top of [PR#56](https://github.com/awslabs/aws-lc/pull/56), this is the third assembly function needed to support the optimized implementation of P-256, nistz256, to be used on ARMv8 as it is currently used only with x86_64.
- This function is used in ECDSA verification; it performs modular inversion in non-constant time since it operates on public values.
- In this PR, pointer authentication was added to this function as well as to `ecp_nistz256_select_w5` and `ecp_nistz256_select_w7`  in the same way as found in most functions in `p256-armv8-asm.pl`  as per OpenSSL.
  A simple overview of the pointer authentication feature in ARMv8.3 can be found in [this presentation by Mark Rutland from ARM](https://events.static.linuxfound.org/sites/events/files/slides/slides_23.pdf).
For more information and references, see [this commit to BoringSSL](https://boringssl-review.googlesource.com/c/boringssl/+/42084) which, when merged, will change PAuth instructions to macros defined in that commit.

### Call-outs:
* This PR would be merged on top of [PR#56](https://github.com/awslabs/aws-lc/pull/56).
* C code duplication continued in this change. This is only transient so that the assembly functions are added and tested gradually. The code duplication will be removed in subsequent PR(s).

### Testing:
* The test for that function was enabled in `p256-x86_64_test.cc` by this change.
* Tested on M6g (ARMv8) and M5a (x86_64) instances, as well as on a MacBook (x86_64).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: patches/0013-p256-armv8-optimizations [hunk 1]
```diff
@@ -2429,20 +2457,460 @@ Index: aws-lc/third_party/boringssl/crypto/fipsmodule/ec/p256-x86_64_test.cc
...
 +#endif
+   bssl::UniquePtr<EC_GROUP> group(
+       EC_GROUP_new_by_curve_name(NID_X9_62_prime256v1));
+   ASSERT_TRUE(group);
+Index: aws-lc/third_party/boringssl/crypto/fipsmodule/ec/asm/p256_beeu-armv8-asm.pl
+===================================================================
+--- /dev/null
++++ aws-lc/third_party/boringssl/crypto/fipsmodule/ec/asm/p256_beeu-armv8-asm.pl
+@@ -0,0 +1,423 @@
++# Copyright (c) 2020, Amazon Inc.
```
Comment (torben-hansen) [inline patches/0013-p256-armv8-optimizations:2494]: There should probably be an explicit license here.

File: patches/0013-p256-armv8-optimizations [hunk 2]
```diff
@@ -2429,20 +2457,460 @@ Index: aws-lc/third_party/boringssl/crypto/fipsmodule/ec/p256-x86_64_test.cc
...
++
++$0 =~ m/(.*[\/\\])[^\/\\]+$/; $dir=$1;
++( $xlate="${dir}arm-xlate.pl" and -f $xlate ) or
++( $xlate="${dir}../../../perlasm/arm-xlate.pl" and -f $xlate) or
++die "can't locate arm-xlate.pl";
++
++open OUT,"| \"$^X\" $xlate $flavour $output";
++*STDOUT=*OUT;
++#############################################################################
++# extern int beeu_mod_inverse_vartime(BN_ULONG out[P256_LIMBS],
```
Comment (torben-hansen) [inline patches/0013-p256-armv8-optimizations:2524]: I see the assumption below (`n` is odd). But is there somewhere else you can document that this algorithm is only good if `n` is odd (if not documented already). Not many would actually read the asm.

File: patches/0013-p256-armv8-optimizations [hunk 3]
```diff
@@ -593,6 +595,19 @@ Index: aws-lc/third_party/boringssl/crypto/fipsmodule/ec/p256.c
 +  }
 +}
 +
++static int ecp_nistz256_scalar_to_montgomery_inv_vartime(const EC_GROUP *group,
++                                                 EC_SCALAR *out,
++                                                 const EC_SCALAR *in) {
++  assert(group->order.width == P256_LIMBS);
```
Comment (torben-hansen) [inline patches/0013-p256-armv8-optimizations:601]: should this be a hard check and not just a debug assert? Related to question above.

File: patches/0013-p256-armv8-optimizations [hunk 4]
```diff
@@ -593,6 +595,19 @@ Index: aws-lc/third_party/boringssl/crypto/fipsmodule/ec/p256.c
 +  }
 +}
 +
++static int ecp_nistz256_scalar_to_montgomery_inv_vartime(const EC_GROUP *group,
```
Comment (torben-hansen) [inline patches/0013-p256-armv8-optimizations:598]: I don't think we have settled how deep in the code we should check for NULL pointers. What do you think?

File: patches/0013-p256-armv8-optimizations [hunk 5]
```diff
@@ -2260,6 +2275,10 @@ Index: aws-lc/third_party/boringssl/crypto/fipsmodule/ec/asm/p256-armv8-asm.pl
 +.type	ecp_nistz256_select_w5,%function
 +.align	4
 +ecp_nistz256_select_w5:
++	.inst	0xd503233f		// paciasp
```
Comment (torben-hansen) [inline patches/0013-p256-armv8-optimizations:2278]: Could you please explain these changes and the other changes made to the functions `ecp_nistz256_select_w5` and `ecp_nistz256_select_w5`?

File: patches/0013-p256-armv8-optimizations [hunk 6]
```diff
@@ -2429,20 +2457,460 @@ Index: aws-lc/third_party/boringssl/crypto/fipsmodule/ec/p256-x86_64_test.cc
...
++  __attribute__((aligned(64))) P256_POINT_AFFINE table[64];
+   for (size_t i = 0; i < 64; i++) {
+     OPENSSL_memset(table[i].X, 2 * i, sizeof(table[i].X));
+     OPENSSL_memset(table[i].Y, 2 * i + 1, sizeof(table[i].Y));
+@@ -99,11 +98,12 @@ TEST(P256_X86_64Test, SelectW7) {
  }
  
-+#if !defined(OPENSSL_AARCH64)
  TEST(P256_X86_64Test, BEEU) {
++#if !defined(OPENSSL_AARCH64)
```
Comment (torben-hansen) [inline patches/0013-p256-armv8-optimizations:2479]: Why is this not relevant for ARM?

File: patches/0013-p256-armv8-optimizations [hunk 7]
```diff
@@ -593,6 +595,19 @@ Index: aws-lc/third_party/boringssl/crypto/fipsmodule/ec/p256.c
 +  }
 +}
 +
++static int ecp_nistz256_scalar_to_montgomery_inv_vartime(const EC_GROUP *group,
++                                                 EC_SCALAR *out,
++                                                 const EC_SCALAR *in) {
++  assert(group->order.width == P256_LIMBS);
++  if (!beeu_mod_inverse_vartime(out->words, in->words, group->order.d)) {
```
Comment (torben-hansen) [inline patches/0013-p256-armv8-optimizations:602]: Can we be explicit here? I.e.  `FAIL_VALUE == beeu_mod_inverse_vartime(out->words, in->words, group->order.d)`

That is more readable.

File: patches/0013-p256-armv8-optimizations [hunk 8]
```diff
@@ -593,6 +595,19 @@ Index: aws-lc/third_party/boringssl/crypto/fipsmodule/ec/p256.c
...
++    return 0;
++  }
++
++  // The result should be returned in the Montgomery domain.
```
Comment (torben-hansen) [inline patches/0013-p256-armv8-optimizations:606]: Not sure what this comment means?

File: patches/0013-p256-armv8-optimizations [hunk 9]
```diff
@@ -2429,20 +2457,460 @@ Index: aws-lc/third_party/boringssl/crypto/fipsmodule/ec/p256-x86_64_test.cc
      !defined(OPENSSL_SMALL) && !defined(BORINGSSL_SHARED_LIBRARY)
  
  TEST(P256_X86_64Test, SelectW5) {
-@@ -98,6 +97,7 @@ TEST(P256_X86_64Test, SelectW7) {
-   CHECK_ABI(ecp_nistz256_select_w7, &val, table, 42);
+   // Fill a table with some garbage input.
+-  alignas(64) P256_POINT table[16];
++  __attribute__((aligned(64))) P256_POINT table[16];
```
Comment (torben-hansen) [inline patches/0013-p256-armv8-optimizations:2462]: Why is this necessary?

File: patches/0013-p256-armv8-optimizations [hunk 10]
```diff
@@ -2429,20 +2457,460 @@ Index: aws-lc/third_party/boringssl/crypto/fipsmodule/ec/p256-x86_64_test.cc
...
++#
++# (Binary Extended GCD (Euclidean) Algorithm.
++#  See A. Menezes, P. vanOorschot, and S. Vanstone's Handbook of Applied Cryptography,
++#  Chapter 14, Algorithm 14.61 and Note 14.64
++#  http://cacr.uwaterloo.ca/hac/about/chap14.pdf)
++#
++# Assumption 1: n is odd for the BEEU
++# Assumption 2: 1 < a < n < 2^256
++
++# Same algorithm in Python:
```
Comment (torben-hansen) [inline patches/0013-p256-armv8-optimizations:2536]: I like this. Thanks.

**Review by torben-hansen — COMMENTED** [review]
Some initial comments. Looking at the ASM next.

File: patches/0013-p256-armv8-optimizations (hunk 1 above)
Comment (nebeid) [reply patches/0013-p256-armv8-optimizations]: I changed it. Thanks.

File: patches/0013-p256-armv8-optimizations (hunk 2 above)
Comment (nebeid) [reply patches/0013-p256-armv8-optimizations:2510]: It's documented with its declaration in `p256-x86_64.h`

File: patches/0013-p256-armv8-optimizations (hunk 3 above)
Comment (nebeid) [reply patches/0013-p256-armv8-optimizations:601]: This comments and other ones are on code that was copied from `p256-x86_64.h`. In this branch https://github.com/nebeid/aws-lc/tree/CA591-clean, the top commit is to remove the copied code and make the x86_64 code shared with AArch64. There will be only one change that ifdef out a variable that doesn't apply to AArch64 (ARMv8) (you asked about it in the test, I'll highlight it below).
So I won't change this code here and we can open stories to improve it.

File: patches/0013-p256-armv8-optimizations (hunk 4 above)
Comment (nebeid) [reply patches/0013-p256-armv8-optimizations:598]: Same comment as https://github.com/awslabs/aws-lc/pull/57/#discussion_r516216015

File: patches/0013-p256-armv8-optimizations (hunk 5 above)
Comment (nebeid) [reply patches/0013-p256-armv8-optimizations:2278]: I explained them in the PR message. They match the rest of the code that was imported from OpenSSL. BoringSSL inserts them differently https://boringssl-review.googlesource.com/c/boringssl/+/42084
and we can change them if we merge that change.

File: patches/0013-p256-armv8-optimizations (hunk 6 above)
Comment (nebeid) [reply patches/0013-p256-armv8-optimizations:2479]: This is what I alluded to in the response to the comment on l.601. The variable underneath this line is only defined for x86_64. In https://github.com/nebeid/aws-lc/tree/CA591-clean, I made it
`#if defined(OPENSSL_X86_64)`
I changed it here too.

File: patches/0013-p256-armv8-optimizations (hunk 7 above)
Comment (nebeid) [reply patches/0013-p256-armv8-optimizations:602]: Same comment as above (l.601)

File: patches/0013-p256-armv8-optimizations (hunk 8 above)
Comment (nebeid) [reply patches/0013-p256-armv8-optimizations:606]: Same comment as above (l.601)

File: patches/0013-p256-armv8-optimizations (hunk 9 above)
Comment (nebeid) [reply patches/0013-p256-armv8-optimizations:2462]: The change of attribute was needed to fix the failing tests of Linux on ARM due to an older version of gcc, older than 9.3, that doesn't have the check on alignas() in C++ code removed.
As for the alignment itself, there is this explanation in `ecp_nistz256_windowed_mul()` in `p256-x86_64.c`:
```
// A |P256_POINT| is (3 * 32) = 96 bytes, and the 64-byte alignment should
+  // add no more than 63 bytes of overhead. Thus, |table| should require
+  // ~1599 ((96 * 16) + 63) bytes of stack space.
+  alignas(64) P256_POINT table[16];
```

File: patches/0013-p256-armv8-optimizations (hunk 10 above)
Comment (nebeid) [reply patches/0013-p256-armv8-optimizations:2522]: Thank you :)

**Review by bryce-shang — APPROVED** [review]

File: patches/0013-p256-armv8-optimizations (hunk 3 above)
Comment (torben-hansen) [reply patches/0013-p256-armv8-optimizations:601]: I see, it is existing code. Fair :)

File: patches/0013-p256-armv8-optimizations (hunk 5 above)
Comment (torben-hansen) [reply patches/0013-p256-armv8-optimizations:2278]: If you could elaborate on how it works?
Not saying it is wrong, just haven't seen it before.

File: patches/0013-p256-armv8-optimizations (hunk 9 above)
Comment (torben-hansen) [reply patches/0013-p256-armv8-optimizations:2462]: Thanks.

File: patches/0013-p256-armv8-optimizations (hunk 2 above)
Comment (torben-hansen) [reply patches/0013-p256-armv8-optimizations:2510]: Ahh I see here: https://github.com/awslabs/aws-lc/blob/main/crypto/fipsmodule/ec/p256-x86_64.h#L96

File: patches/0013-p256-armv8-optimizations (hunk 5 above)
Comment (nebeid) [reply patches/0013-p256-armv8-optimizations:2278]: That's fair. I edited the PR message. Please let me know if it is what you're hoping to see.

**Review by torben-hansen — COMMENTED** [review]
Made it until `Lbeeu_loop_endz`.

Publishing now to do things in parallel.

File: patches/0013-p256-armv8-optimizations [hunk 11]
```diff
@@ -2429,20 +2457,446 @@ Index: aws-lc/third_party/boringssl/crypto/fipsmodule/ec/p256-x86_64_test.cc
...
++    // then out = -Y mod n
++
++    // Verify that A = 1 ==> (-1)*Y*a = A = 1  (mod |n|)
++    // Is A-1 == 0?
++    // If not, fail.
++    sub     $t0, $a0, #1
++    orr     $t0, $t0, $a1
++    orr     $t0, $t0, $a2
++    orr     $t0, $t0, $a3
++    cbnz    $t0, .Lbeeu_err
```
Comment (torben-hansen) [inline patches/0013-p256-armv8-optimizations:2839]: Do we have test cases hitting this condition?

File: patches/0013-p256-armv8-optimizations [hunk 12]
```diff
@@ -2429,20 +2457,446 @@ Index: aws-lc/third_party/boringssl/crypto/fipsmodule/ec/p256-x86_64_test.cc
...
++
++    // If Y>n ==> Y:=Y-n
++.Lbeeu_reduction_loop:
++    // x_i := y_i - n_i (X is no longer needed, use it as temp)
++    // ($t0 = 0 from above)
++    subs    $x0, $y0, $n0
++    sbcs    $x1, $y1, $n1
++    sbcs    $x2, $y2, $n2
++    sbcs    $x3, $y3, $n3
++    sbcs    $x4, $y4, $t0
```
Comment (torben-hansen) [inline patches/0013-p256-armv8-optimizations:2849]: There is no fancy way to skip this when no borrow?

**Review by torben-hansen — COMMENTED** [review]
Last comments.

Good job on documenting the assembly. Helped me a lot.

File: patches/0013-p256-armv8-optimizations [hunk 13]
```diff
@@ -2429,20 +2457,446 @@ Index: aws-lc/third_party/boringssl/crypto/fipsmodule/ec/p256-x86_64_test.cc
...
++#
++# (Binary Extended GCD (Euclidean) Algorithm.
++#  See A. Menezes, P. vanOorschot, and S. Vanstone's Handbook of Applied Cryptography,
++#  Chapter 14, Algorithm 14.61 and Note 14.64
++#  http://cacr.uwaterloo.ca/hac/about/chap14.pdf)
++#
++# Assumption 1: n is odd for the BEEU
++# Assumption 2: 1 < a < n < 2^256
++
++# Same algorithm in Python:
```
[the remark this answers is not in the document]
Comment (nebeid) [reply patches/0013-p256-armv8-optimizations:2522]: I added explanation to address these 3 points. I'm not if it's overcommenting, but a reader can skip the details.

File: patches/0013-p256-armv8-optimizations [hunk 14]
```diff
@@ -2429,20 +2457,446 @@ Index: aws-lc/third_party/boringssl/crypto/fipsmodule/ec/p256-x86_64_test.cc
...
++# x0-x2, x30 are used to hold the modulus n. The input parameters passed in
++# x1,x2 are copied first before corrupting them. x0 (out) is stored on the stack.
++# x3-x7 are used for parameters, which is not the case in this function, so they are corruptible
++# x8 is corruptible here
++# (the function doesn't return a struct, hence x8 doesn't contain a passed-in address
++#  for that struct).
++# x9-x15 are corruptible registers
++# x19-x28 are callee-saved registers
++
++# X/Y will hold the inverse parameter
```
[the remark this answers is not in the document]
Comment (nebeid) [reply patches/0013-p256-armv8-optimizations:2564]: Many of these comments are inherited from `p256_beeu-x86_64-asm.pl` to maintain the correspondence between the two files if a reader would like to check that this algorithm behaves like its x86_64 counterpart.
In this case, as I understand it, it means that
```
    Xa == B (mod n)
   -Ya == A (mod n)
```
are maintained.

File: patches/0013-p256-armv8-optimizations [hunk 15]
```diff
@@ -2429,20 +2457,446 @@ Index: aws-lc/third_party/boringssl/crypto/fipsmodule/ec/p256-x86_64_test.cc
...
++#     add n and goto shift1_<ctr>
++# Prerequisite: t0 = 0
++$g_next_label = 0;
++sub SHIFT1 {
++  my ($var0, $var1, $var2, $var3, $var4) = @_;
++  my $label = ".Lshift1_${g_next_label}";
++  $g_next_label++;
++  return <<___;
++    tbz     $var0, #0, $label
++    adds    $var0, $var0, $n0
```
[the remark this answers is not in the document]
Comment (nebeid) [reply patches/0013-p256-armv8-optimizations:2607]: In general, you wouldn't want to add the carry in the least significant word unless you've explicitly set it to 0. Instead of doing that, you can simply ignore it.

File: patches/0013-p256-armv8-optimizations [hunk 16]
```diff
@@ -2429,20 +2457,446 @@ Index: aws-lc/third_party/boringssl/crypto/fipsmodule/ec/p256-x86_64_test.cc
...
++# "uint() mod 64" is from the definition of LSL and LSR instructions.
++#
++# What matters here is the order of instructions relative to certain other
++# instructions, i.e.
++# - lsr and lsl must precede orr of the corresponding registers.
++# - lsl must preced the lsr of the same register afterwards.
++# The chosen order of the instructions overall is to try and maximize
++# the pipeline usage.
++sub SHIFT256 {
++  my ($var0, $var1, $var2, $var3, $var4) = @_;
```
[the remark this answers is not in the document]
Comment (nebeid) [reply patches/0013-p256-armv8-optimizations:2648]: Good catch. Thank you. Fixed.

File: patches/0013-p256-armv8-optimizations [hunk 17]
```diff
@@ -2429,20 +2457,446 @@ Index: aws-lc/third_party/boringssl/crypto/fipsmodule/ec/p256-x86_64_test.cc
...
++    eor     $y2, $y2, $y2
++    eor     $y3, $y3, $y3
++    eor     $y4, $y4, $y4
++
++.Lbeeu_loop:
++    // if B == 0, jump to .Lbeeu_loop_end
++    ${\TEST_B_ZERO}
++
++    // 0 < B < |n|,
++    // 0 < A <= |n|,
```
[the remark this answers is not in the document]
Comment (nebeid) [reply patches/0013-p256-armv8-optimizations:2731]: A is only as big as _n_ at the beginning and then A and B keep decreasing. _n_ fits in four 64-bit registers based on the initial assumptions.

File: patches/0013-p256-armv8-optimizations [hunk 18]
```diff
@@ -2429,20 +2457,446 @@ Index: aws-lc/third_party/boringssl/crypto/fipsmodule/ec/p256-x86_64_test.cc
...
++
++    // shift := number of trailing 0s in $b0
++    // (      = number of leading 0s in $t1; see the "rbit" instruction in TEST_B_ZERO)
++    clz     $shift, $t1
++
++    // If there is no shift, goto shift_A_Y
++    cbz     $shift, .Lbeeu_shift_A_Y
++
++    // Shift B right by "$shift" bits
++    ${\SHIFT256($b0, $b1, $b2, $b3)}
```
[the remark this answers is not in the document]
Comment (nebeid) [reply patches/0013-p256-armv8-optimizations:2747]: Thanks. Wasn't my idea, it's coming from `p256_beeu-x86_64-asm.pl`. Counting the number of trailing 0s is different here in order to save few instructions and jumps.

File: patches/0013-p256-armv8-optimizations [hunk 19]
```diff
@@ -2429,20 +2457,446 @@ Index: aws-lc/third_party/boringssl/crypto/fipsmodule/ec/p256-x86_64_test.cc
...
++    // $shift--;
++    // $t0 := 0; needed in the addition to the most significant word in SHIFT1
++    eor     $t0, $t0, $t0
++.Lbeeu_shift_loop_Y:
++    ${\SHIFT1($y0, $y1, $y2, $y3, $y4)}
++    subs    $shift, $shift, #1
++    bne     .Lbeeu_shift_loop_Y
++
++.Lbeeu_update_B_X_or_A_Y:
++    // Try T := B - A; if cs, continue with B > A (cs: carry set = no borrow)
```
[the remark this answers is not in the document]
Comment (nebeid) [reply patches/0013-p256-armv8-optimizations:2790]: That's a good question. I understand if you try to find the inverse of small values like 1 and 2, T will be ~_n_, just slightly smaller.

File: patches/0013-p256-armv8-optimizations (hunk 11 above)
Comment (nebeid) [reply patches/0013-p256-armv8-optimizations:2839]: I didn't test the algorithm with _n_ being other than the group order of curve p256. So all numbers modulo n have an inverse and hence, A, the gcd, would be 1. All the tests in `p256-x86_64_test.cc` are also based on that modulus. 
You're right that this means this code may not be covered for now, but it's not used in the caller for other use cases than the group order of P256. We can create a story to cover this part of the code.

File: patches/0013-p256-armv8-optimizations (hunk 12 above)
Comment (nebeid) [reply patches/0013-p256-armv8-optimizations:2849]: If I understood your question correctly, I think you wonder if you can compare quickly instead of going through a full multi-precision subtraction. 
I thought about that and found that comparing will need to compare corresponding words starting from the most significant, but there are cases, you would go down to less significant words if the most significant ones are equal (e.g., zeros), then you need to select the words of the larger value to be the minuend and the other one the subtrahend. So in terms of code size, it may be larger, and the effort may not be saved, since the number of instructions may increase.
In this architecture, the pipeline can execute more instructions per clock cycle, and in the worst case, if gated on the carry, it will be one instruction per cycle.

File: patches/0013-p256-armv8-optimizations [hunk 20]
```diff
@@ -2429,20 +2457,446 @@ Index: aws-lc/third_party/boringssl/crypto/fipsmodule/ec/p256-x86_64_test.cc
...
++    subs    $t0, $b0, $a0
```
[the remark this answers is not in the document]
Comment (nebeid) [reply patches/0013-p256-armv8-optimizations:2791]: As in the previous comment, with _a_ = 1 (or 2, or any small number). Initially, A = _n_ and B= _a_. So B-A always becomes negative and the next set of instructions gets executed. These test cases are included in the `for` loop of `TEST(P256_X86_64Test, BEEU)` in `p256-x86_64_test.cc`. 
In fact, I debugged this code by running the case of _a_ = 2 through the python code and this code after extracting the result from the test. It takes many iterations for the gap between B and A to become smaller.

File: patches/0013-p256-armv8-optimizations (hunk 13 above)
[the remark this answers is not in the document]
Comment (torben-hansen) [reply patches/0013-p256-armv8-optimizations:2546]: This explains the invariant equations. Nice.

File: patches/0013-p256-armv8-optimizations (hunk 14 above)
[the remark this answers is not in the document]
Comment (torben-hansen) [reply patches/0013-p256-armv8-optimizations:2589]: Ok. Just not sure how to interpret "inverse parameter". NP, never mind :)

File: patches/0013-p256-armv8-optimizations (hunk 17 above)
[the remark this answers is not in the document]
Comment (torben-hansen) [reply patches/0013-p256-armv8-optimizations:2756]: Gotcha `Assumption: a,n,X,Y < 2^(256)`.

File: patches/0013-p256-armv8-optimizations (hunk 19 above)
[the remark this answers is not in the document]
Comment (torben-hansen) [reply patches/0013-p256-armv8-optimizations:2815]: My followup question is then whether this fit in four 64-bit registers taking sign into account?

File: patches/0013-p256-armv8-optimizations (hunk 20 above)
[the remark this answers is not in the document]
Comment (torben-hansen) [reply patches/0013-p256-armv8-optimizations:2816]: Good good. I just wanted to make sure there are tests that takes a path where you borrow, don't borrow and approaches the maximum/minimal values. If the possible input values allow it.

File: patches/0013-p256-armv8-optimizations (hunk 11 above)
Comment (torben-hansen) [reply patches/0013-p256-armv8-optimizations:2864]: I see. How much work to add a test with some other input? It doesn't not have to be large input.

File: patches/0013-p256-armv8-optimizations (hunk 12 above)
Comment (torben-hansen) [reply patches/0013-p256-armv8-optimizations:2874]: In the last `sbcs` instruction the subtrahend is `t0 = 0`. My assumption was that you do this to take any borrow into account. NP

File: patches/0013-p256-armv8-optimizations (hunk 19 above)
[the remark this answers is not in the document]
Comment (nebeid) [reply patches/0013-p256-armv8-optimizations:2815]: That's a very good question. I think the table on this page https://community.arm.com/developer/ip-products/processors/b/processors-ip-blog/posts/condition-codes-1-condition-flags-and-codes
explains the condition codes well.
For example, let _unsigned_ B and A be
```
B = 1 = 0x0001
A = 15 = 0x1111
Then, 
B - A = 0x0001 + 0x0001
```
There is no space for the sign bit of A, but the processor's ALU, Arithmetic Logic Unit, treats both signed and unsigned numbers the same.
The result will be `0x0010` (which, if there were extra sign bits, should be `0x11110010 = -0x00001110 = -14`).
Our indication that the result is negative is that the carry bit was not set.
If you do 
```
A - B = 0x1111 + 0x1111
```
There will be a carry, which means no-borrow, because A > B.
Note: If we were doing _signed_ arithmetic, i.e. we had the most significant bit as the sign bit, we would need to check the overflow bit as in the table on that page.

File: patches/0013-p256-armv8-optimizations (hunk 11 above)
Comment (nebeid) [reply patches/0013-p256-armv8-optimizations:2864]: I think it's not much work. We can set n to 9 and a to 3 and test that there is no inverse, because gcd(3,9) = 3.
I'd prefer to do it in another story as I would add to the tests in `p256-x86_64_test.cc`.
For the changes in this patch, I'm maintaining the original x86_64 C code and tests.

File: patches/0013-p256-armv8-optimizations (hunk 12 above)
Comment (nebeid) [reply patches/0013-p256-armv8-optimizations:2874]: Correct.

File: patches/0013-p256-armv8-optimizations (hunk 11 above)
Comment (torben-hansen) [reply patches/0013-p256-armv8-optimizations:2864]: Since this is not going to main, I am good with that.

File: patches/0013-p256-armv8-optimizations (hunk 20 above)
[the remark this answers is not in the document]
Comment (torben-hansen) [reply patches/0013-p256-armv8-optimizations:2816]: We can add this to a new story as the similar comment below.

File: patches/0013-p256-armv8-optimizations (hunk 19 above)
[the remark this answers is not in the document]
Comment (torben-hansen) [reply patches/0013-p256-armv8-optimizations:2815]: A 1-2 sentence with this analysis/argument would be great to add here for future reference.

[the remark this answers is not in the document]
Comment (nebeid) [reply patches/0013-p256-armv8-optimizations:2815]: I expanded the comments on this section as per your recommendation.

**Review by torben-hansen — APPROVED** [review]
Nice!

File: patches/0013-p256-armv8-optimizations (hunk 11 above)
Comment (nebeid) [reply patches/0013-p256-armv8-optimizations:2868]: I missed to mention there is one test case in `p256-x86_64_test.cc` that exercises this condition which is _a_ = 0.
